src/groupeddatatable/grouping.jl

```diff
@@ -117,49 +117,38 @@ df |> groupby([:a, :b]) |> [sum, length]
 ```
 
 """
-function groupby{T}(d::AbstractDataTable, cols::Vector{T})
-    ## a subset of Wes McKinney's algorithm here:
-    ##     http://wesmckinney.com/blog/?p=489
-
-    ncols = length(cols)
-    # use CategoricalArray to get a set of integer references for each unique item
-    nv = NullableCategoricalArray(d[cols[ncols]])
-    # if there are NULLs, add 1 to the refs to avoid underflows in x later
-    anynulls = (findfirst(nv.refs, 0) > 0 ? 1 : 0)
-    # use UInt32 instead of the original array's integer size since the number of levels can be high
-    x = similar(nv.refs, UInt32)
-    for i = 1:nrow(d)
-        if nv.refs[i] == 0
-            x[i] = 1
-        else
-            x[i] = CategoricalArrays.order(nv.pool)[nv.refs[i]] + anynulls
+function groupby(d::AbstractDataTable, cols::Vector)
+    intersect = d[cols]
+    mappings = OrderedDict{DataTableRow, Vector{Int}}()
+    for i = 1:nrow(intersect)
+        row = DataTableRow(intersect, i)
+        if !haskey(mappings, row)
+            mappings[row] = [i]
+        elseif haskey(mappings, row)
+            push!(mappings[row], i)
         end
     end
-    # also compute the number of groups, which is the product of the set lengths
-    ngroups = length(levels(nv)) + anynulls
-    # if there's more than 1 column, do roughly the same thing repeatedly
-    for j = (ncols - 1):-1:1
-        nv = NullableCategoricalArray(d[cols[j]])
-        anynulls = (findfirst(nv.refs, 0) > 0 ? 1 : 0)
-        for i = 1:nrow(d)
-            if nv.refs[i] != 0
-                x[i] += (CategoricalArrays.order(nv.pool)[nv.refs[i]] + anynulls - 1) * ngroups
-            end
-        end
-        ngroups = ngroups * (length(levels(nv)) + anynulls)
-        # TODO if ngroups is really big, shrink it
+    ngroups = length(mappings.keys)
+    idx = Vector{Int}(nrow(d))
+    starts = fill(1, ngroups)
+    stops = Vector{Int}(ngroups)
+
+    rows = mappings.vals[1]
+    idx[1:length(rows)] = rows
+    stops[1] = length(rows)
+    for i = 2:ngroups
+        rows = mappings.vals[i]
+        starts[i] = stops[i-1] + 1
+        stops[i] = stops[i-1] + length(rows)
+        idx[starts[i]:stops[i]] = rows
     end
-    (idx, starts) = groupsort_indexer(x, ngroups)
```

Comment:
This function is also used in [`join`](https://github.com/JuliaData/DataTables.jl/blob/ee17d92e69eca03744320a065caaf8f50e255ad1/src/abstractdatatable/join.jl#L23). Should we move the function to that file since it's no longer used here?